Just a few changes, needing reviewed.

contificate opened the pull request:
It's just a few small changes I thought I could commit. 

Some of the changes might break the intention you guys intended so this PR is mostly a review.

EDIT: I had to amend this PR because I forgot to commit a header file, hopefully the AppVeyor and Travis CI builds will go through now. EDIT: Just noticed some of the formatting's off, my bad.

contificate committed:
Removing superfluous empty constructors/deconstructors.

The declaration of these is superfluous as special member generation will implicitly create them anyway for classes whose members are considered trivial (have well defined default constructors).
If you're gonna implement other constructors in the future (for example: I believe it'd be clearer to just create a factory that uses dependency injection via non-default constructors would be better than allowing the client to create default - empty - rules, for example), the edited classes will be deemed non-trivial and the default constructor will be implicitly deleted. In a case like that, just default them so that the compiler continues to generate them (Class() = default;).

I believe this commit is reasonable considering the ctors/dtors aren't virtual.

contificate committed:
Removing duplicated code from immutable overloads of const methods.
Passing std::string by const reference.
Sets default parameter for getMatch() to 0 such that getFirstMatch just falls through to that default.

contificate committed:
Make the const accessor method for YaraRule return a const reference to a std::string.
That way the user can choose to invoke a copy construction if they choose.

Attempted to clarify logic (you might wanna go back and edit the yara_detector.cpp if you think my edit is non-obvious).
Implemented an output stream overload for writing YaraRule's name.
Some other small formatting changes.

contificate committed:
Including file forgotten in previous commit.

metthal commented on include/yaracpp/types/yara_rule.h:
```diff
@@ -41,21 +38,26 @@ class YaraRule
 		void addMatch(const YaraMatch &match);
 		/// @}
+
+        // @name Utility operator overloads
```
Rest of the file uses tabs so please replace these spaces with tabs.

metthal commented on src/types/yara_rule.cpp:
```diff
@@ -57,9 +41,9 @@ const YaraMeta* YaraRule::getMeta(std::string id) const
  * @return Pointer to selected match or @c nullptr if such match is not found
  */
-const YaraMatch* YaraRule::getMatch(std::size_t index) const
+const YaraMatch* YaraRule::getMatch(std::size_t index = {}) const
```
If you want to add default parameter to this method do so in header file. This way, only this file would be able to see this default parameter but you would not be able to use it outside this compile unit. Also reconsider whether this change is really needed. There is already method `getFirstMatch()` so I don't really see this change to be that useful.

metthal commented on src/yara_detector/yara_detector.cpp:
```diff
@@ -245,11 +245,8 @@ int YaraDetector::yaraCallback(int message, void *messageData, void *userData)
-
-	return result == 0;
+	const auto result = yr_compiler_add_string(compiler, string, nullptr);
+	return (needsRecompilation = !(result));
```
I would prefer readability over compactness. Even though your code got rid of the branches, when I see this I am not sure whether the author intended the assignment or he just mistyped equal to.

metthal commented:
Thanks for the first pull request! I went through your changes and made some comments. I agree with most of the changes and would be happy to merge them. I just think that few small things need to be addressed first.

s3rvac commented on include/yaracpp/types/yara_rule.h:
```diff
@@ -41,21 +38,26 @@ class YaraRule
+
+        // @name Utility operator overloads
+        // @{
+        friend std::ostream& operator<<(std::ostream& o, const YaraRule& rule);
```
I would add `#include <iosfwd>` into the header file for `std::ostream&` because otherwise, you might get `error: ‘ostream’ in namespace ‘std’ does not name a type`.

Also, I would add `#include <ostream>` into the `.cpp` file, as you call `std::ostream::operator<<()` in there, which requires a full definition of `std::ostream`.

contificate committed:
Modified files as requested by maintainers.
Briefly:
- Clarified logic in YaraDetector::addRules
- Amended formatting (hopefully - it'd be nice if the projects shipped with a .clangformat; without being pedantic, the source looks similar to LLVM style so I guess I could run ClangFormat with that style and see how it goes)
- Added includes for the std::ostream << operator overload for YaraRule.
- Removed default value in YaraRule::getMatch(). My intention was to perhaps - later on - provide some kind of iterable interface for accessing the rules, without relying on accessors to the private STL container that's used to store the rules. I didn't wanna pollute the intention of the maintainers since I think other Avast projects rely on this repo as a dependency.

obscurecolin commented:
I made the requested changes.

Just waiting on AppVeyor to go through. Travis CI is having issues with its OSX target so I expect it to be "Pending" indefinitely.

metthal commented:
Sorry I am getting back to you after such a long time. There were just lot of things I needed to take care of and I simply forgot about this PR. Thanks for the contribution. Merging.
